Libraries/LMCommon/LanguageModel.swift

```diff
@@ -0,0 +1,160 @@
+// Copyright © 2024 Apple Inc.
+
+import Foundation
+@preconcurrency import Hub
+import MLX
+import MLXNN
+import Tokenizers
+
+extension Module {
+
+    /// Compute the number of parameters in a possibly quantized model
+    public func numParameters() -> Int {
+        return leafModules().flattenedValues().map {
+            mod -> Int in
+            if let qlin = mod as? QuantizedLinear {
+                return qlin.scales.size * qlin.groupSize
+            } else if let qemb = mod as? QuantizedEmbedding {
+                return qemb.scales.size * qemb.groupSize
+            } else {
+                return mod.parameters().flattenedValues().reduce(
+                    0,
+                    {
+                        $0 + $1.size
+                    })
+            }
+        }.reduce(0, +)
+    }
+}
+
+public struct LMInput {
+    public let text: Text
+    public let image: ProcessedImage?
+
+    public struct Text {
+        public let tokens: MLXArray
+        public let mask: MLXArray?
+
+        public init(tokens: MLXArray, mask: MLXArray? = nil) {
+            self.tokens = tokens
+            self.mask = mask
+        }
+
+        public subscript(
+            indices: MLXArrayIndex..., stream stream: StreamOrDevice = .default
+        ) -> Text {
+            Text(tokens: tokens[indices, stream: stream], mask: mask?[indices, stream: stream])
+        }
+
+        public subscript(
+            text indices: MLXArrayIndex..., stream stream: StreamOrDevice = .default
+        ) -> Text {
+            Text(tokens: tokens[indices, stream: stream], mask: mask)
+        }
+    }
+
+    public struct ProcessedImage {
+        public let pixels: MLXArray
+        public let imageSizes: MLXArray?
+        public let imageGridThw: MLXArray?
+        public let aspectRatioIds: MLXArray?
+        public let aspectRatioMask: MLXArray?
+
+        public init(
+            pixels: MLXArray, imageSizes: MLXArray? = nil, imageGridThw: MLXArray? = nil,
+            aspectRatioIds: MLXArray? = nil, aspectRatioMask: MLXArray? = nil
+        ) {
+            self.pixels = pixels
+            self.imageSizes = imageSizes
+            self.imageGridThw = imageGridThw
+            self.aspectRatioIds = aspectRatioIds
+            self.aspectRatioMask = aspectRatioMask
+        }
+    }
+
+    public init(tokens: MLXArray, mask: MLXArray? = nil) {
+        self.init(text: .init(tokens: tokens, mask: mask))
+    }
+
+    public init(text: LMInput.Text, image: LMInput.ProcessedImage? = nil) {
+        self.text = text
+        self.image = image
+    }
+}
+
+public struct LMOutput {
```

Comment:
Union type for the output.  Some of the VLMs return additional state, which is represented here.